### DISCUSSION TITLE
Prevent space after inline headings

### DISCUSSION THREAD

**Discussion by Karmac** [typst/typst#1173] [Q&A] [answered]
Hello,

I'm using inline headings. The problem is typst inserts a blank space after any line break, which breaks my layout.
Let me show you an example:

```typ
#show heading.where(level: 2): it => {
  box(it.body + [.---])
}

== Heading 1
#lorem(20)

#heading(level: 2)[Heading 2]#lorem(20)
```

This prints the following:

![image](https://github.com/typst/typst/assets/638179/253cf097-4f40-4248-992e-b3c43c422ff5)

As you can see, the line break in the first heading results in a blank space after it.
Even though the second approach works as intended, I'd still like to use the == syntax for headings because it is more legible.

So, I've tried regex rules, but they apply after the content has been parsed so there's no way to select a heading. Queries don't work either because they don't let you remove content. Using `#h(-0.25em)` seems like a workaround, but when using `justify: true` the width of the space varies, leading to unintended results.

I've come to this solution:
```typ
show: rest => {
  let child = rest.children
  let i = 0
  
  for item in child {     
    if item.func() == heading and item.level >= 2 {
        child.remove(i + 1)
    } else {
      i += 1
    }
  }

  // HOW TO PRINT THE MODIFIED DOCUMENT?
  rest
}
```

But I need help to render `child` again. Any other idea on how to implement this?

Thanks.

**Comment by Julian-Wassmann** [comment, marked as answer]
Your approach with the rest show rule works in general. I modified it just a bit.

This will do the job:
```
// heading show rule needs to be before rest show rule
#show heading.where(level: 2): it => {
  box(it.body + [.---])
}

// rest show rule needs to be after heading show rule and before document content
#show: rest => {
  for (n, item) in rest.children.enumerate() {
    // skip first item, so that lastItem is always correct
    if n == 0 { continue }
    // get the previous item
    let lastItem = rest.children.at(n - 1)
    // if current item is a whitespace and heading is above, then...
    if item == [ ] and lastItem.func() == heading {
      // ...remove the whitespace
      item = []
    }
    // return the (modified) item
    item
  }
}

// This works now
== Heading 1
#lorem(20)

// For testing: This whitespace is not removed because no heading is above
-#[ ]-

// This still works
#heading(level: 2)[Heading 2]#lorem(20)

// This also works
#heading(level: 2)[Heading 3] #lorem(20)
```

**Comment by Karmac** [comment]
Awesome, thanks!
